src/components/navigation/SlideMenuContext.tsx

```diff
@@ -0,0 +1,75 @@
+import React, { createContext, useCallback, useContext, useEffect, useMemo, useState } from 'react';
+
+import { useHistory } from 'react-router-dom';
+
+import { SlideMenuCtx } from '@app/types';
+
+import { navigationItemParentMap } from './NavigationItems';
+
+interface ActiveNavItem {
+  id: string;
+  path: string;
+  parents: string[];
+}
+
+const SlideMenuContext = createContext<SlideMenuCtx>(null);
+const useSlideMenuState = (): SlideMenuCtx => useContext(SlideMenuContext);
+
+const SlideMenuProvider: React.FC = props => {
+  const [expandedNavItems, setExpandedNavItems] = useState<string[]>([]);
+  const [activeNavItem, setActiveNavItem] = useState<ActiveNavItem | undefined>();
+  const history = useHistory();
+
+  const addExpandedNavItem = useCallback((id: string): void => {
+    setExpandedNavItems(prevExpandedNavItems => {
+      if (prevExpandedNavItems.includes(id)) {
+        return prevExpandedNavItems;
+      }
+      return [...prevExpandedNavItems, id];
+    });
+  }, []);
+
+  const removeExpandedNavItem = useCallback((id: string): void => {
+    setExpandedNavItems(prevExpandedNavItems => {
+      if (!prevExpandedNavItems.includes(id)) {
+        return prevExpandedNavItems;
+      }
+      return prevExpandedNavItems.filter(item => item !== id);
+    });
+  }, []);
+
+  useEffect(() => {
+    const unlisten = history.listen(location => {
+      const activeItemKey = Object.keys(navigationItemParentMap).find(
+        key => navigationItemParentMap[key].path === location.pathname,
+      );
+      const activeItem = activeItemKey
+        ? {
+            id: activeItemKey,
+            ...navigationItemParentMap[activeItemKey],
+          }
+        : undefined;
+      setExpandedNavItems(activeItem?.parents ? [...activeItem.parents] : []);
+      setActiveNavItem(activeItem);
+    });
+
+    return unlisten;
+  }, [addExpandedNavItem, history]);
+
+  const contextValue = useMemo(
+    () => ({
+      expandedNavItems,
+      activeNavItem,
+      expandNavItem: addExpandedNavItem,
+      collapseNavItem: removeExpandedNavItem,
+    }),
+    [activeNavItem, addExpandedNavItem, expandedNavItems, removeExpandedNavItem],
+  );
```

Comment:
Some of the items in that array are memoized callbacks that will themselves never change. I'd like to keep the context memoized since it holds centralized state for the entire nav menu and if it gets redefined on every render (which could happen outside of a state change too if a parent component re-renders) that will lead to a lot of unnecessary re-renders of components down the tree.

Reply:
I can live with that, better safe than sorry